From 39a41530414ddd20252f4b0393396638943534e0 Mon Sep 17 00:00:00 2001
From: Jouni Malinen <j@w1.fi>
Date: Sun, 1 Apr 2012 17:55:20 +0300
Subject: [PATCH 66/73] Fix memory leaks on radius_client_send error paths

In case this function returns an error, the RADIUS message needs to
freed in the caller.

Signed-hostap: Jouni Malinen <j@w1.fi>
(cherry picked from commit ca8e039fec1352e2ece4fdbd6fde92fd461565a3)
---
 radius_example/radius_example.c |    3 ++-
 src/ap/accounting.c             |   15 +++++++++------
 src/ap/ieee802_11_auth.c        |    3 ++-
 wpa_supplicant/eapol_test.c     |    4 +++-
 4 files changed, 16 insertions(+), 9 deletions(-)

--- a/src/ap/accounting.c
+++ b/src/ap/accounting.c
@@ -265,8 +265,9 @@ void accounting_sta_start(struct hostapd
 			       hapd, sta);
 
 	msg = accounting_msg(hapd, sta, RADIUS_ACCT_STATUS_TYPE_START);
-	if (msg)
-		radius_client_send(hapd->radius, msg, RADIUS_ACCT, sta->addr);
+	if (msg &&
+	    radius_client_send(hapd->radius, msg, RADIUS_ACCT, sta->addr) < 0)
+		radius_msg_free(msg);
 
 	sta->acct_session_started = 1;
 }
@@ -364,9 +365,10 @@ static void accounting_sta_report(struct
 		goto fail;
 	}
 
-	radius_client_send(hapd->radius, msg,
-			   stop ? RADIUS_ACCT : RADIUS_ACCT_INTERIM,
-			   sta->addr);
+	if (radius_client_send(hapd->radius, msg,
+			       stop ? RADIUS_ACCT : RADIUS_ACCT_INTERIM,
+			       sta->addr) < 0)
+		goto fail;
 	return;
 
  fail:
@@ -469,7 +471,8 @@ static void accounting_report_state(stru
 		return;
 	}
 
-	radius_client_send(hapd->radius, msg, RADIUS_ACCT, NULL);
+	if (radius_client_send(hapd->radius, msg, RADIUS_ACCT, NULL) < 0)
+		radius_msg_free(msg);
 }
 
 
--- a/src/ap/ieee802_11_auth.c
+++ b/src/ap/ieee802_11_auth.c
@@ -191,7 +191,8 @@ static int hostapd_radius_acl_query(stru
 		goto fail;
 	}
 
-	radius_client_send(hapd->radius, msg, RADIUS_AUTH, addr);
+	if (radius_client_send(hapd->radius, msg, RADIUS_AUTH, addr) < 0)
+		goto fail;
 	return 0;
 
  fail:
--- a/wpa_supplicant/eapol_test.c
+++ b/wpa_supplicant/eapol_test.c
@@ -284,7 +284,9 @@ static void ieee802_1x_encapsulate_radiu
 		}
 	}
 
-	radius_client_send(e->radius, msg, RADIUS_AUTH, e->wpa_s->own_addr);
+	if (radius_client_send(e->radius, msg, RADIUS_AUTH, e->wpa_s->own_addr)
+	    < 0)
+		goto fail;
 	return;
 
  fail:
